File: src/main/java/io/github/binaryfoo/Decoder.java
package io.github.binaryfoo;

import io.github.binaryfoo.decoders.DecodeSession;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface Decoder {

    /**
     * Turn bits into something more mind friendly.
     */
	public List<DecodedData> decode(@NotNull String input, int startIndexInBytes, @NotNull DecodeSession decodeSession);

    /**
     * Return null if ok. Otherwise an abusive/informative/educational message.
     */
	public String validate(String input);

    /**
     * In characters.
     */
	public int getMaxLength();
}

File: src/main/java/io/github/binaryfoo/Decoder.kt
package io.github.binaryfoo

import io.github.binaryfoo.decoders.DecodeSession

/**
 * Idea being children are shown in a tree structure.
 */
interface Decoder {

    /**
     * Turn bits into something more mind friendly.
     */
    fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): List<DecodedData>

    /**
     * Return null if ok. Otherwise an abusive/informative/educational message.
     */
    fun validate(input: String?): String?

    /**
     * In characters.
     */
    fun getMaxLength(): Int
}
